### PR TITLE
[FLINK-11364][tests] Port TaskManagerFailsITCase to new code base

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

Port `TaskManagerFailsITCase` to new code base.

## Brief change log

- "detect a failing task manager" --> JobMaster#testHeartbeatTimeoutWithTaskManager

- "handle gracefully failing task manager" --> JobMasterTest#testJobFailureWhenGracefulTaskExecutorTermination

- "handle hard failing task manager" --> JobMasterTest#testJobFailureWhenTaskExecutorHeartbeatTimeout

- "go into a clean state in case of a TaskManager failure" --> TaskExecutorITCase#testNewTaskExecutorJoinsCluster

## Verifying this change

- Run ported tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by GJL** [comment]
@flinkbot approve description
@flinkbot approve consensus

File: flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java [hunk 1]
```diff
@@ -66,6 +67,17 @@ public TestingMiniCluster(TestingMiniClusterConfiguration miniClusterConfigurati
...
 	}
 
+	@Nonnull
+	@Override
+	public CompletableFuture<Void> terminateTaskExecutor(int index) {
+		return super.terminateTaskExecutor(index);
+	}
+
+	@Override
+	public void startTaskExecutor(boolean localCommunication) throws Exception {
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java]: Doesn't look like user friendly API. Shouldn't `MiniCluster` set this flag depending on the configuration? Wouldn't it be enough to expose a signature such as `public void startTaskExecutor()`

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		}
+	}
+
+	/**
+	 * Tests that a job will be re-executed if a new TaskExecutor joins the cluster.
+	 */
+	@Test
+	public void testNewTaskExecutorJoinsCluster() throws Exception {
+
+		final Deadline deadline = Deadline.fromNow(TESTING_TIMEOUT);
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java]: Declaration of `deadline` is too early. Should be closer to `waitUntilCondition`

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 3]
```diff
@@ -1430,6 +1430,120 @@ public void testReleasingTaskExecutorIfNoMoreSlotsRegistered() throws Exception
...
+
+			final ExecutionAttemptID executionAttemptId = taskDeploymentFuture.get();
+
+			jobMasterGateway.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), executionAttemptId, ExecutionState.RUNNING)).get();
+
+			jobReachedRunningState.accept(taskManagerLocation, jobMasterGateway);
+
+			final ArchivedExecutionGraph archivedExecutionGraph = onCompletionActions.getJobReachedGloballyTerminalStateFuture().get();
+
+			assertThat(archivedExecutionGraph.getState(), is(JobStatus.FAILED));
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: Method is too long.

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java [hunk 4]
```diff
@@ -696,34 +686,38 @@ protected ResourceManagerRunner startResourceManager(
...
-	protected TaskExecutor[] startTaskManagers(
-			Configuration configuration,
-			HighAvailabilityServices haServices,
-			HeartbeatServices heartbeatServices,
-			MetricRegistry metricRegistry,
-			BlobCacheService blobCacheService,
-			int numTaskManagers,
-			RpcServiceFactory rpcServiceFactory) throws Exception {
+	@GuardedBy("lock")
+	private void startTaskManagers() throws Exception {
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: True. However, I will move it to the lifecycle method section.

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java [hunk 5]
```diff
@@ -738,6 +732,26 @@ protected ResourceManagerRunner startResourceManager(
 	//  Internal methods
 	// ------------------------------------------------------------------------
 
+	@GuardedBy("lock")
+	private Collection<? extends CompletableFuture<Void>> terminateTaskExecutor() {
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: Yes definitely.

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java [hunk 6]
```diff
@@ -945,4 +954,12 @@ public void onFatalError(Throwable exception) {
 			closeAsync();
 		}
 	}
+
+	private class TerminatingFatalErrorHandlerFactory {
+
+		@GuardedBy("lock")
+		private TerminatingFatalErrorHandler create() {
+			return new TerminatingFatalErrorHandler(taskManagers.size());
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: Good point, this is quite strongly coupled. I'll introduce an `int index` parameter which is given to `create`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: I think 50 lines of a testing method is not too long. Happy to refactor it once we see duplicated code in other parts.

File: flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java]: Well, `MiniCluster` cannot know whether to use local communication or not at start time, because `TestingMiniCluster` allows to start new `TaskExecutors`. Thus, the only option would be to always set `localCommunication` to false in the case of the `MiniCluster`.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java]: I introduced a new method `useLocalCommunication` which can be overriden by the `TestingMiniCluster` to always set local communication to `false`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java]: True, will move it closer to the call site.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+		return new JobGraph("Blocking test job", vertex);
+	}
+
+	/**
+	 * Blocking invokable which is controlled by a static field.
+	 */
+	public static class BlockingOperator extends AbstractInvokable {
+		private static final Object lock = new Object();
+		private static volatile boolean isBlocking = true;
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java]: Good point. I'll change the code using the `CountDownLatch` and rename the field to `blocking`.

**Comment by tillrohrmann** [comment]
Thanks for the review @GJL. I've addressed your comments and pushed a fixup.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 8]
```diff
@@ -1430,6 +1422,125 @@ public void testReleasingTaskExecutorIfNoMoreSlotsRegistered() throws Exception
...
+			(jobMasterGateway, taskManagerResourceId) -> resourceId -> {
+				if (respondToHeartbeats.get()) {
+					jobMasterGateway.heartbeatFromTaskManager(taskManagerResourceId, new AccumulatorReport(Collections.emptyList()));
+				}
+			}
+		);
+	}
+
+	private void runJobFailureWhenTaskExecutorTerminatesTest(
+			Supplier<HeartbeatServices> heartbeatSupplier,
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: Does `HeartbeatServices` have to be lazily supplied?

```suggestion
			HeartbeatServices heartbeatServices,
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 9]
```diff
@@ -1430,6 +1422,125 @@ public void testReleasingTaskExecutorIfNoMoreSlotsRegistered() throws Exception
...
+			BiConsumer<LocalTaskManagerLocation, JobMasterGateway> jobReachedRunningState,
+			BiFunction<JobMasterGateway, ResourceID, Consumer<ResourceID>> heartbeatConsumerFunction) throws Exception {
+		final JobGraph jobGraph = createSingleVertexJobGraph();
+		final TestingOnCompletionActions onCompletionActions = new TestingOnCompletionActions();
+		final JobMaster jobMaster = createJobMaster(
+			new Configuration(),
+			jobGraph,
+			haServices,
+			new TestingJobManagerSharedServicesBuilder().build(),
+			heartbeatSupplier.get(),
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]:
```suggestion
			heartbeatServices,
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 10]
```diff
@@ -1430,6 +1422,125 @@ public void testReleasingTaskExecutorIfNoMoreSlotsRegistered() throws Exception
...
+				localTaskManagerLocation.getResourceID(),
+				new FlinkException("Test disconnectTaskManager exception.")),
+			(jobMasterGateway, resourceID) -> ignored -> {});
+	}
+
+	@Test
+	public void testJobFailureWhenTaskExecutorHeartbeatTimeout() throws Exception {
+		final AtomicBoolean respondToHeartbeats = new AtomicBoolean(true);
+		runJobFailureWhenTaskExecutorTerminatesTest(
+			() -> fastHeartbeatServices,
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]:
```suggestion
			fastHeartbeatServices,
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 11]
```diff
@@ -1430,6 +1422,125 @@ public void testReleasingTaskExecutorIfNoMoreSlotsRegistered() throws Exception
...
+	}
+
+	/**
+	 * Tests that the job execution is failed if the TaskExecutor disconnects from the
+	 * JobMaster.
+	 */
+	@Test
+	public void testJobFailureWhenGracefulTaskExecutorTermination() throws Exception {
+		runJobFailureWhenTaskExecutorTerminatesTest(
+			() -> heartbeatServices,
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]:
```suggestion
			heartbeatServices,
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java (hunk 3 above)
Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: If you deem it readable, leave it as is.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 12]
```diff
@@ -1430,6 +1422,125 @@ public void testReleasingTaskExecutorIfNoMoreSlotsRegistered() throws Exception
...
+		final SlotOffer slotOffer = new SlotOffer(new AllocationID(), 0, ResourceProfile.UNKNOWN);
+		final Collection<SlotOffer> slotOffers = jobMasterGateway.offerSlots(taskManagerLocation.getResourceID(), Collections.singleton(slotOffer), testingTimeout).get();
+
+		assertThat(slotOffers, hasSize(1));
+	}
+
+	private static final class TestingOnCompletionActions implements OnCompletionActions {
+
+		private final CompletableFuture<ArchivedExecutionGraph> jobReachedGloballyTerminalStateFuture = new CompletableFuture<>();
+		private final CompletableFuture<Void> jobFinishedByOtherFuture = new CompletableFuture<>();
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: This unused field is for future extensions?

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java [hunk 13]
```diff
@@ -928,9 +931,16 @@ public void onFatalError(Throwable exception) {
 
 	private class TerminatingFatalErrorHandlerFactory {
 
+		/**
+		 * Create a new {@link TerminatingFatalErrorHandler} for the {@link TaskExecutor} with
+		 * the given index.
+		 *
+		 * @param index into the {{@link #taskManagers}} collection to identify the correct {@link TaskExecutor}.
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: Are the double curly braces `{{}}` intended?

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java [hunk 14]
```diff
@@ -215,6 +215,14 @@ public HighAvailabilityServices getHighAvailabilityServices() {
 		}
 	}
 
+	@VisibleForTesting
+	@Nonnull
+	protected Collection<DispatcherResourceManagerComponent<?>> getDispatcherResourceManagerComponents() {
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: I am confused about why this change is show in the fixup commit.

File: flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java [hunk 15]
```diff
@@ -74,8 +74,13 @@ public TestingMiniCluster(TestingMiniClusterConfiguration miniClusterConfigurati
 	}
 
 	@Override
-	public void startTaskExecutor(boolean localCommunication) throws Exception {
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java]: Is it crucial for testing to be able to set the right `localCommunication` flag? If yes, a method overload that sets `localCommunication` to `false` would have been enough but I am not insisting on it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java]: ok

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (GJL) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: ok

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (GJL) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: ok

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (GJL) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: ok

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java (hunk 12 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: yes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java (hunk 13 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: unintended. will change it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java (hunk 8 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: No, this is not correct. Will change.

File: flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java (hunk 15 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java]: are you suggesting to always start the `TaskExecutors` with `localCommunication = false`? Or only for the `TestingMiniCluster`? The latter should now be the case.

**Comment by tillrohrmann** [comment]
Thanks for the second round of review @GJL. I addressed your comments except for the last one for which I didn't understand your proposal yet.

**Comment by GJL** [comment]
@flinkbot approve all

File: flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java (hunk 15 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/minicluster/TestingMiniCluster.java]: Actually, this functionality was needed for #7690.

**Comment by tillrohrmann** [comment]
Thanks for the review @GJL. Merging once Travis gives green light.
